src/hooks/useSignChallenge.ts

```diff
@@ -0,0 +1,123 @@
+import { useEffect, useState, useCallback, useRef } from 'react';
+import { useSignMessage } from 'wagmi';
+import { SIGNING_SERVICE_URL } from '../constants/constants';
+import { storageKeys } from '../constants/localStorage';
+import { SiweMessage } from 'siwe';
+
+export interface SiweSignatureData {
+  nonce: string;
+  signature: string;
+  expirationDate: string;
+}
+
+export function useSiweSignature(address?: `0x${string}`) {
+  const { signMessageAsync } = useSignMessage();
+  const [requiresSign, setRequiresSign] = useState(false);
+
+  // Used to wait for the modal interaction and/or return of the
+  // signing promise.
+  const signPromiseRef = useRef<{
```

Comment:
This one was interesting. At first I tried that, but since `checkAndWaitForSignature` and the `force` counterpart were being used in the `sep-10` function which is not a hook, I was experiencing issues and the state was not reflecting the change.

I later changed the implementation (before I was polling for the signature with a `setTimeout`) so it might be that there is no issue storing the promise in a simple state. I would have to test again.